### PR TITLE
DEPR: Enforce deprecation of partial failure in transform for lists/dicts

### PR DESCRIPTION
**Pull request by rhshadrach**
From #43740

- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/apply/test_frame_transform.py [hunk 1]
```diff
@@ -133,32 +133,35 @@ def func(x):
...
         )
 
     obj = DataFrame({"A": 3 * [object]})  # DataFrame that will fail on most transforms
     obj = tm.get_obj(obj, frame_or_series)
+    if op == "rank":
+        error = ValueError
+        msg = "Transform function failed"
+    else:
+        error = TypeError
+        msg = (
```
Comment (jbrockmendel) [inline pandas/tests/apply/test_frame_transform.py:148]: nitpick can you use the `msg = "|".join([...])` pattern here

**Comment by jbrockmendel** [comment]
small nitpick, otherwise LGTM.  excited to see this change

File: pandas/tests/apply/test_frame_transform.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/tests/apply/test_frame_transform.py]: Thanks - agreed that's much easier to read.

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jbrockmendel** [comment]
thanks @rhshadrach pls be sure to update #30228

**Comment by rhshadrach** [comment]
It's on the same line as #49401, so will update after that PR is merged.

**Pull request merged**